pkg/errlist/errlist.go

```diff
@@ -6,35 +6,132 @@ import (
 	"go/token"
 	"io"
 	"path/filepath"
+	"sort"
 	"strings"
+
+	"encr.dev/pkg/errinsrc"
+	"encr.dev/pkg/errinsrc/srcerrors"
 )
 
+// Verbose controls whether the error list prints all errors
+// or just the first two.
+var Verbose = false
+
+// MaxErrorsToPrint is the maximum number of errors to print
+// if Verbose is false
+var MaxErrorsToPrint = 1
+
 type List struct {
-	list scanner.ErrorList
+	List errinsrc.List `json:"list,omitempty"`
 	fset *token.FileSet
 }
 
+var _ errinsrc.ErrorList = (*List)(nil)
+
 func New(fset *token.FileSet) *List {
 	return &List{fset: fset}
 }
 
+// Convert attempts to convert known error types into an error list
+// if it can't it returns nil
+func Convert(err error) *List {
+	switch err := err.(type) {
+	case *List:
+		return err
+	case *errinsrc.ErrInSrc:
+		l := New(nil)
+		l.List = []*errinsrc.ErrInSrc{err}
+		return l
+	case errinsrc.ErrorList:
+		l := New(nil)
+		l.List = err.ErrorList()
+		return l
+	default:
+		return nil
+	}
+}
+
+// Report is a function that allows you to report an error into
+// this list, without having to check for nil.
+//
+// This function only supports error of types:
+//   - *List
+//   - *errinsrc.ErrInSrc
+//   - *scanner.Error
+//
+// If too many errors have been reported it panics
+// with a Bailout value to abort processing.
+// Use HandleBailout to conveniently handle this.
+func (l *List) Report(err error) {
+	if err == nil {
+		return
+	}
+
+	var errToAdd *errinsrc.ErrInSrc
+	switch err := (err).(type) {
+	case *errinsrc.ErrInSrc:
+		// the base type we expect
+		errToAdd = err
+
+	case *scanner.Error:
+		// errors directly from the Go parser
+		errToAdd = srcerrors.GenericGoParserError(err)
+
+	case scanner.ErrorList:
+		for _, e := range err {
+			l.Report(e)
+		}
+		return
+	case *List:
+		// If it's a different list, then merge it in
+		if err != l {
+			for _, e := range err.ErrorList() {
+				l.Report(e)
+			}
+		}
+	case errinsrc.ErrorList:
+		// either another errlist or a list from `srcerrors`
+		for _, e := range err.ErrorList() {
+			l.Report(e)
+		}
+		return
+	default:
+		panic(fmt.Sprintf("unsupported type %T being reported to errlist.List", err))
+	}
+
+	// Skip adding this error if it's on the same line as another error
+	// we've already reported, since it's probably a spurious error caused
+	// by the first one.
+	for _, e := range l.List {
+		if errToAdd.OnSameLine(e) {
```

Comment:
We don't, but this is behaviour that the old `errlist` had, so I wanted to maintain that behaviour rather than change it.

i.e. the first error reported on each line was kept, other errors where dropped.

Do you think we should change that behaviour?

Reply:
how well does it work in practice? How common is it with multiple errors per line?

Reply:
No idea to be honest. @eandre thoughts?